AnyText/AnyText.Core/Parser.FoldingRanges.cs

```diff
@@ -0,0 +1,180 @@
+﻿using NMF.AnyText.Model;
+using NMF.AnyText.Rules;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.XPath;
+
+namespace NMF.AnyText
+{
+    public partial class Parser
+    {
+        public ParsedFoldingRange[] GetFoldingRangesFromRoot()
+        {
+            RuleApplication rootApplication = Context.RootRuleApplication;
+
+            if (rootApplication.IsPositive)
+            {
+                var parsedFoldingRanges = GetFoldingRangesFromSequence(rootApplication);
+                return parsedFoldingRanges;
+            }
+
+            return Array.Empty<ParsedFoldingRange>();
+        }
+
+        private ParsedFoldingRange[] GetFoldingRangesFromSequence(RuleApplication ruleApplication)
+        {
+            var result = new List<ParsedFoldingRange>();
+
+            if (ruleApplication.Comments != null)
+            {
+                result.AddRange(GetCommentFoldingRanges(ruleApplication));
+            }
+
+            if (ruleApplication.Rule is ZeroOrMoreRule zeroOrMoreRule && zeroOrMoreRule.InnerRule.IsImports())
+            {
+                result.Add(GetImportsFoldingRange(ruleApplication));
+            }
+
+            if (ruleApplication is MultiRuleApplication multiRuleApplication)
+            {
+                if (multiRuleApplication.Rule is SequenceRule sequenceRule)
+                {
+                    if (sequenceRule.IsRegion())
+                    {
+                        result.Add(GetRegionFoldingRange(multiRuleApplication));
+                    }
+                    else if (multiRuleApplication.Rule is ParanthesesRule || sequenceRule.IsFoldable())
+                    {
+                        result.Add(GetFoldingRange(multiRuleApplication));
+                    }
+                }
+
+                foreach (var innerRuleApplication in multiRuleApplication.Inner)
+                {
+                    var innerFoldingRanges = GetFoldingRangesFromSequence(innerRuleApplication);
+                    result.AddRange(innerFoldingRanges);
```

Comment:
ich fände es schöner, wenn wir eine Liste rumreichen würden in die alle reinschreiben, anstatt ständig eine Liste in ein Array zu kopieren, um das Array dann wieder in eine Liste kopieren zu können